File: php-laravel/app/Http/Requests/UpdateToolRequest.php
<?php

namespace App\Http\Requests;

use Illuminate\Foundation\Http\FormRequest;

class UpdateToolRequest extends FormRequest
{
    /**
     * Determine if the user is authorized to make this request.
     */
    public function authorize(): bool
    {
        return true;
    }

    /**
     * Get the validation rules that apply to the request.
     */
    public function rules(): array
    {
        $toolId = $this->route('tool');
        
        return [
            'name' => 'sometimes|string|min:2|max:100|unique:tools,name,' . $toolId,
            'description' => 'nullable|string',
            'vendor' => 'sometimes|string|max:100',
            'website_url' => 'nullable|url|max:255',
            'category_id' => 'sometimes|integer|exists:categories,id',
            'monthly_cost' => 'sometimes|numeric|min:0|decimal:0,2',
            'active_users_count' => 'nullable|integer|min:0',
            'owner_department' => 'sometimes|in:Engineering,Sales,Marketing,HR,Finance,Operations,Design',
            'status' => 'sometimes|in:active,deprecated,trial'
        ];
    }

    /**
     * Get custom error messages for validation rules.
     */
    public function messages(): array
    {
        return [
            'name.unique' => 'A tool with this name already exists',
            'name.min' => 'Tool name must be at least 2 characters',
            'name.max' => 'Tool name must not exceed 100 characters',
            'category_id.exists' => 'Selected category does not exist',
            'monthly_cost.min' => 'Monthly cost must be a positive number',
            'monthly_cost.decimal' => 'Monthly cost must have at most 2 decimal places',
            'owner_department.in' => 'Invalid department. Must be one of: Engineering, Sales, Marketing, HR, Finance, Operations, Design',
            'website_url.url' => 'Website URL must be a valid URL format'
        ];
    }
}

File: java-springboot/src/main/java/com/techcorp/internaltools/dto/UpdateToolRequest.java
package com.techcorp.internaltools.dto;

import com.techcorp.internaltools.model.Department;
import com.techcorp.internaltools.model.ToolStatus;
import io.swagger.v3.oas.annotations.media.Schema;
import jakarta.validation.constraints.*;
import lombok.Data;

import java.math.BigDecimal;

@Data
@Schema(description = "Request to update an existing tool")
public class UpdateToolRequest {
    
    @Size(min = 2, max = 100, message = "Name must be between 2 and 100 characters")
    @Schema(description = "Tool name", example = "Slack")
    private String name;

    @Schema(description = "Tool description", example = "Updated team messaging platform")
    private String description;

    @Size(max = 100, message = "Vendor name must not exceed 100 characters")
    @Schema(description = "Vendor/provider name", example = "Slack Technologies")
    private String vendor;

    @Pattern(regexp = "^https?://.*", message = "Must be a valid URL format")
    @Schema(description = "Tool website URL", example = "https://slack.com")
    private String websiteUrl;

    @Schema(description = "Category ID", example = "1")
    private Long categoryId;

    @DecimalMin(value = "0.0", message = "Must be a positive number")
    @Digits(integer = 10, fraction = 2, message = "Must have maximum 2 decimal places")
    @Schema(description = "Monthly cost per user", example = "9.50")
    private BigDecimal monthlyCost;

    @Schema(description = "Owner department", example = "Engineering")
    private Department ownerDepartment;

    @Schema(description = "Tool status", example = "active")
    private ToolStatus status;

    @Min(value = 0, message = "Active users count must be non-negative")
    @Schema(description = "Number of active users", example = "25")
    private Integer activeUsersCount;
}
